### PR TITLE
Add gathering facts for HC correctly installed

### LINKED ISSUE #101: [HyperShift] Gather must-gather from each Hosted Cluster
We need a must-gather from each Hosted Cluster after the churn workload finishes and before the Hosted Clusters are deleted to debug any problems identified through dashboards later. Collecting so many must-gathers on the jump host might not be possible - so we may need to send them over to some nfs share.

### PR DESCRIPTION
**Pull request by morenod**
Adding two functions for getting information from hosted clusters:
- _get_must_gather, using the oc adm must gather
- _get_dump_cluster, using the hypershift dump cluster-

Commented the _get_must_gather execution as it is not available at this moment

Files will be saved inside each hosted cluster folder

### CODE REVIEW

File: hypershift/hosted-wrapper.py [hunk 1]
```diff
@@ -198,8 +213,12 @@ def _build_cluster(hypershift_cmnd, kubeconfig_location, cluster_name_seed, mgmt
                 logging.info('Waiting for all clusters to be installed to start e2e-benchmarking execution on %s' % cluster_name)
                 all_clusters_installed.wait()
             logging.info('Executing e2e-benchmarking to add load on the cluster %s with %s nodes during %s with %d iterations' % (cluster_name, str(worker_nodes), load_duration, job_iterations))
-            _cluster_load(kubeconfig_location, my_path, cluster_name, load_duration, job_iterations, es_url, mgmt_cluster_name)
+            _cluster_load(my_path, cluster_name, load_duration, job_iterations, es_url, mgmt_cluster_name)
             logging.info('Finished execution of e2e-benchmarking workload on %s' % cluster_name)
+        #logging.info("Saving must-gather file of hosted cluster %s" % cluster_name)
```
Comment (dry923) [inline hypershift/hosted-wrapper.py]: Can we get rid of the old commands if we aren't using them here?

**Comment by mukrishn** [comment]
LGTM just a few suggestions,
1. Can we have it optional(thru argument) also to collect only on failed nodes to save some space? 
2. Can we move the files to snappy server using [this](https://github.com/cloud-bulldozer/e2e-benchmarking/tree/master/utils/scale-ci-diagnosis) existing script ? (can be a separate PR)

**Comment by morenod** [comment]
> LGTM just a few suggestions,
> 
> 1. Can we have it optional(thru argument) also to collect only on failed nodes to save some space?
> 2. Can we move the files to snappy server using [this](https://github.com/cloud-bulldozer/e2e-benchmarking/tree/master/utils/scale-ci-diagnosis) existing script ? (can be a separate PR)

Related to 1. could we execute dump from failed nodes? 
With current code, must_gather/dump will be executed only when hypershift create cluster ends with exit code = 0. From my point of view, it makes no sense to execute must gather on a failed installation, because, in most cases, we wont have access to it. if we want to identify an installation failure, we already have the installation log.

Related to 2, sending files out of the server is out of scope of this script, we would need to deal with connection failures, authentication and other infrastructure problems not related with the purpose of the script. If we want to have the dump on other server, lets create another tool for it

**Comment by mukrishn** [comment]
> Related to 1.  Dump logs from failed cluster. 

IMO collection of logs when exit 1 is must and optional on exit 0 often we won't bother.
`hypershift create dump` can be executed on a failed or partial state hostedcluster since its a cli execution, an active mgmt cluster with failed hostedcluster resource is enough to collect this and access to hosted cluster isn't required. Also, during an occurrence of a failure hypershift team needs that dump to troubleshoot further along with the CLI error message. Also, tomorrow with rosa cli we must collect rosa logs when there is a failure so always good to collect log when there is exit 1. You're right about must-gather as it need cluster access and its not possible on hypershift cluster today. 

> Related to 2, sending files out of the server is out of scope of this script

Okay, if this script is not intended to do that part we can omit it.

**Comment by morenod** [comment]
> > Related to 1.  Dump logs from failed cluster.
> 
> IMO collection of logs when exit 1 is must and optional on exit 0 often we won't bother. `hypershift create dump` can be executed on a failed or partial state hostedcluster since its a cli execution, an active mgmt cluster with failed hostedcluster resource is enough to collect this and access to hosted cluster isn't required. Also, during an occurrence of a failure hypershift team needs that dump to troubleshoot further along with the CLI error message. Also, tomorrow with rosa cli we must collect rosa logs when there is a failure so always good to collect log when there is exit 1. You're right about must-gather as it need cluster access and its not possible on hypershift cluster today.
> 
> > Related to 2, sending files out of the server is out of scope of this script
> 
> Okay, if this script is not intended to do that part we can omit it.

Added parameter to only collect must gather from failed clusters or collect all.
@dry923  Added also must-gather for management cluster

**Comment by morenod** [comment]
Added all fixes included in the PR #103 here.

- max(0,random) on the jobs calculation

Also added a random sleep before dumping must-gather of hosted cluster

File: hypershift/hosted-wrapper.py [hunk 2]
```diff
@@ -259,6 +267,97 @@ def _cluster_load(kubeconfig, my_path, hosted_cluster_name, load_duration, jobs,
...
+        logging.error("Failed to compress must-gather of %s cluster from %s to %s" % (cluster_name, cluster_path + "/must_gather", cluster_path + "must_gather.tar.gz"))
+        return 1
+    logging.info('Deleting non-compressed must-gather files of hosted cluster %s' % cluster_name)
+    must_gather_delete_command = ["rm", "-rf", cluster_path + "/must_gather"]
+    logging.debug(must_gather_delete_command)
+    must_gather_delete_process = subprocess.Popen(must_gather_delete_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, env=myenv)
+    must_gather_delete_stdout, must_gather_delete_stderr = must_gather_delete_process.communicate()
+    if must_gather_delete_process.returncode != 0:
+        logging.error("Failed to delete non-compressed must-gather files of hosted cluster %s" % cluster_name)
+        return 1
```
Comment (dry923) [inline hypershift/hosted-wrapper.py:298]: I don't think we should "fail" the test b/c we couldnt get a must gather as the actual workload may have run fine. Can we change these to logging.warn and just alert that it didnt work in the logs? Same for all the other ones that do a must gather

Comment (morenod) [reply hypershift/hosted-wrapper.py:298]: It wont fail the test, that return is not used to determine the status. I need to return something, to stop the execution of the function, so it wont compress a file that not exists or something like that

I can change to return 0 and logging.warning, but that wont affect to the final result of the script

Comment (dry923) [reply hypershift/hosted-wrapper.py:298]: ack. disregard then

**Review by dry923 — APPROVED** [review]
LGTM!

**Review by mukrishn — APPROVED** [review]
LGTM!!

**Pull request merged**